### PR TITLE
feat: Added Vote Locking

### PR DESCRIPTION
**Pull request by hexonaut**
Added the ability for governance to set a vote locking duration which will only allow calling the `free()` function after the duration has elapsed when voting. Takes into account delegate voting by checking your delegate as well. If you switch your delegate after they vote, the vote locking period will be transferred to your user.

`unlockTime` was split into `voteUnlockTime` and `proposalUnlockTime` because I didn't want voting to block the ability to add another proposal.

### CODE REVIEW

File: src/DssGov.sol [hunk 1]
```diff
@@ -487,6 +501,9 @@ contract DssGov {
         // Verify amount for voting is lower or equal than voting rights
         require(wad <= _getUserRights(msg.sender, snapshotIndex, proposals[id].blockNum), "DssGov/amount-exceeds-rights");
 
+        // Update locked time
+        users[msg.sender].voteUnlockTime = _max(users[msg.sender].voteUnlockTime, _add(block.timestamp, voteLockDuration));
```
Comment (gbalabasquer) [inline src/DssGov.sol:505]: Is the `_max` actually necessary? I understand the case of a malicious proposal setting it to 0 and overwriting current locked users. However if a malicious proposal made to pass much worse things for sure already happened.

**Comment by gbalabasquer** [comment]
Btw pushed a merge commit from master to solve a conflict after the change we had discussed.

File: src/DssGov.sol (hunk 1 above)
Comment (hexonaut) [reply src/DssGov.sol:505]: You are correct in that I was thinking of the case where a malicious proposal sets the lock to 0. This is a good point that if a malicious proposal passes this is a worst case scenario, and anything is possible. Locking the MKR won't matter at that point. I think you are right - it's probably better to save the gas.

**Review by gbalabasquer — APPROVED** [review]
LGTM

**Pull request merged**